ksql-common/src/main/java/io/confluent/ksql/util/SchemaUtil.java

```diff
@@ -348,24 +349,57 @@ public static boolean areCompatible(final Schema arg1, final Schema arg2) {
   }
 
   private static boolean mapEquals(final Schema mapA, final Schema mapB) {
-    return Objects.equals(mapA.keySchema(), mapB.keySchema())
-        && Objects.equals(mapA.valueSchema(), mapB.valueSchema());
+    return areCompatible(mapA.keySchema(), mapB.keySchema())
+        && areCompatible(mapA.valueSchema(), mapB.valueSchema());
   }
 
   private static boolean arrayEquals(final Schema arrayA, final Schema arrayB) {
-    return Objects.equals(arrayA.valueSchema(), arrayB.valueSchema());
+    return areCompatible(arrayA.valueSchema(), arrayB.valueSchema());
   }
 
   private static boolean structEquals(final Schema structA, final Schema structB) {
     return structA.fields().isEmpty()
         || structB.fields().isEmpty()
-        || Objects.equals(structA.fields(), structB.fields());
+        || compareFieldsOfStructs(structA, structB);
```

Comment:
Not particularly pertinent to your change, but do you know why a two structs are considered equal if one side has no fields?

Maybe this has something to do with the semantics being "are they compatible" and not "are they equal"...?

As above, if that's the case I think we should rename these methods

Reply:
Hmm, good question. I don't know but I will discuss with Almog today and try to find out why.